File: src/avenue_mcp/rag/extract.py
"""Text extraction with position information.

Positions become citations, so every extractor records where text came from:
page number for PDF, slide number for PPTX, heading path for DOCX/HTML.

Scanned PDFs are a real failure case -- some course outlines are photocopies.
These are detected and reported rather than silently indexed as empty
documents, because a file that appears indexed but never retrieves leaves the
user wondering why searching for their outline finds nothing.
"""

from __future__ import annotations

import hashlib
import logging
from dataclasses import dataclass, field
from pathlib import Path

from avenue_mcp.errors import ExtractionError
from avenue_mcp.util.html import to_text

log = logging.getLogger(__name__)

# Below this many characters across a multi-page document, assume the text
# layer is missing (scanned image).
_POOR_TEXT_THRESHOLD = 100

SUPPORTED_SUFFIXES = {
    ".pdf", ".docx", ".pptx", ".txt", ".md", ".markdown",
    ".html", ".htm", ".csv", ".log", ".rst",
}


@dataclass
class Segment:
    """A positioned run of text."""

    text: str
    position: str          # "p.3" | "slide 14" | "§ Grading" | "line 1"
    heading: str | None = None
    page: int | None = None


@dataclass
class Extraction:
    segments: list[Segment] = field(default_factory=list)
    page_count: int | None = None
    quality: str = "ok"     # ok | poor | empty
    note: str | None = None

    @property
    def full_text(self) -> str:
        return "\n\n".join(s.text for s in self.segments if s.text.strip())

    @property
    def char_count(self) -> int:
        return sum(len(s.text) for s in self.segments)


def is_supported(path: Path | str) -> bool:
    return Path(path).suffix.lower() in SUPPORTED_SUFFIXES


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with Path(path).open("rb") as fh:
        for block in iter(lambda: fh.read(131072), b""):
            h.update(block)
    return h.hexdigest()


def extract(path: Path | str, mime_type: str | None = None) -> Extraction:
    """Dispatch on suffix, falling back to MIME type."""
    p = Path(path)
    if not p.is_file():
        raise ExtractionError(f"No such file: {p}")

    suffix = p.suffix.lower()
    if not suffix and mime_type:
        suffix = {
            "application/pdf": ".pdf",
            "text/plain": ".txt",
            "text/html": ".html",
        }.get(mime_type.split(";")[0].strip(), "")

    try:
        if suffix == ".pdf":
            result = _pdf(p)
        elif suffix == ".docx":
            result = _docx(p)
        elif suffix == ".pptx":
            result = _pptx(p)
        elif suffix in (".html", ".htm"):
            result = _html(p)
        elif suffix in (".txt", ".md", ".markdown", ".csv", ".log", ".rst"):
            result = _plain(p)
        else:
            raise ExtractionError(f"Unsupported file type: {suffix or 'unknown'}")
    except ExtractionError:
        raise
    except Exception as exc:  # noqa: BLE001 -- one bad file must not kill a sync
        raise ExtractionError(f"Could not read {p.name}: {exc}") from exc

    return _assess(result)


def _assess(result: Extraction) -> Extraction:
    chars = result.char_count
    if chars == 0:
        result.quality = "empty"
        result.note = "No text could be extracted."
    elif chars < _POOR_TEXT_THRESHOLD and (result.page_count or 1) > 1:
        result.quality = "poor"
        result.note = (
            "Almost no text extracted from a multi-page document -- this is "
            "likely a scanned image. OCR is not supported."
        )
    return result


# --- PDF ------------------------------------------------------------------


def _pdf(path: Path) -> Extraction:
    try:
        import pymupdf  # type: ignore
    except ImportError:  # pragma: no cover
        try:
            import fitz as pymupdf  # type: ignore
        except ImportError as exc:
            raise ExtractionError("pymupdf is not installed") from exc

    segments: list[Segment] = []
    with pymupdf.open(path) as doc:
        pages = doc.page_count
        for i, page in enumerate(doc, start=1):
            text = (page.get_text("text") or "").strip()
            if text:
                segments.append(Segment(text=text, position=f"p.{i}", page=i))
    return Extraction(segments=segments, page_count=pages)


# --- DOCX -----------------------------------------------------------------


def _docx(path: Path) -> Extraction:
    try:
        import docx  # type: ignore
    except ImportError as exc:  # pragma: no cover
        raise ExtractionError("python-docx is not installed") from exc

    document = docx.Document(str(path))
    segments: list[Segment] = []
    heading = None
    buffer: list[str] = []

    def flush() -> None:
        if buffer:
            body = "\n".join(buffer).strip()
            if body:
                segments.append(
                    Segment(
                        text=body,
                        position=f"§ {heading}" if heading else "§ body",
                        heading=heading,
                    )
                )
            buffer.clear()

    for para in document.paragraphs:
        text = (para.text or "").strip()
        style = (para.style.name or "").lower() if para.style is not None else ""
        if style.startswith("heading") and text:
            flush()
            heading = text
            buffer.append(text)
        elif text:
            buffer.append(text)
    flush()

    for ti, table in enumerate(document.tables, start=1):
        rows = []
        for row in table.rows:
            cells = [(c.text or "").strip() for c in row.cells]
            if any(cells):
                rows.append(" | ".join(cells))
        if rows:
            segments.append(
                Segment(text="\n".join(rows), position=f"table {ti}", heading=heading)
            )

    return Extraction(segments=segments)


# --- PPTX -----------------------------------------------------------------


def _pptx(path: Path) -> Extraction:
    try:
        from pptx import Presentation  # type: ignore
    except ImportError as exc:  # pragma: no cover
        raise ExtractionError("python-pptx is not installed") from exc

    prs = Presentation(str(path))
    segments: list[Segment] = []
    slides = 0

    for idx, slide in enumerate(prs.slides, start=1):
        slides = idx
        parts: list[str] = []
        for shape in slide.shapes:
            if getattr(shape, "has_text_frame", False):
                t = (shape.text_frame.text or "").strip()
                if t:
                    parts.append(t)
            if getattr(shape, "has_table", False):
                for row in shape.table.rows:
                    cells = [(c.text or "").strip() for c in row.cells]
                    if any(cells):
                        parts.append(" | ".join(cells))

        # Speaker notes often carry the actual explanation -- the slide says
        # "Amortized Analysis" and the notes explain it.
        try:
            if slide.has_notes_slide:
                notes = (slide.notes_slide.notes_text_frame.text or "").strip()
                if notes:
                    parts.append(f"[Speaker notes] {notes}")
        except Exception:  # noqa: BLE001
            pass

        if parts:
            segments.append(
                Segment(text="\n".join(parts), position=f"slide {idx}", page=idx)
            )

    return Extraction(segments=segments, page_count=slides)


# --- HTML / plain ---------------------------------------------------------


def _html(path: Path) -> Extraction:
    raw = path.read_text("utf-8", errors="replace")
    text = to_text(raw)
    segments = (
        [Segment(text=text, position="§ document")] if text.strip() else []
    )
    return Extraction(segments=segments)


def _plain(path: Path) -> Extraction:
    text = path.read_text("utf-8", errors="replace").strip()
    if not text:
        return Extraction(segments=[])
    # Split on blank lines so long text files still chunk sensibly.
    blocks = [b.strip() for b in text.split("\n\n") if b.strip()]
    segments = [
        Segment(text=b, position=f"block {i}") for i, b in enumerate(blocks, start=1)
    ]
    return Extraction(segments=segments)

File: extension/src/rag/extract.ts
/**
 * Text extraction, with position information — because positions become
 * citations, and an unattributed passage about a late penalty is worse than no
 * answer: the student cannot check it, and if retrieval was wrong they act on
 * a policy from a different course.
 *
 * | Format   | Library            | Position unit |
 * |----------|--------------------|---------------|
 * | PDF      | pdf.js             | page number   |
 * | PPTX     | JSZip + DOMParser  | slide number  |
 * | DOCX     | JSZip + DOMParser  | paragraph run |
 * | HTML/TXT | DOMParser / raw    | section       |
 *
 * SCANNED PDFs ARE A REAL FAILURE CASE. Some course outlines are photocopies,
 * and pdf.js returns near-empty text for them. Detected and reported rather
 * than silently indexed: a document that appears indexed but never retrieves
 * leaves the user wondering why searching for their outline finds nothing.
 * OCR is out of scope.
 */

import JSZip from "jszip";

export interface Segment {
  /** "p.3" / "slide 14" / "§ Grading" */
  position: string;
  text: string;
  /** Slide notes and headings are worth keeping distinct while chunking. */
  kind?: "body" | "notes" | "heading";
}

export interface Extraction {
  segments: Segment[];
  pageCount: number | null;
  quality: "ok" | "poor";
  /** Populated when quality is "poor", so the sync can report a real reason. */
  reason?: string;
}

export class ExtractionError extends Error {}

const SUPPORTED = [".pdf", ".pptx", ".docx", ".html", ".htm", ".txt", ".md", ".csv"];

export function isSupported(fileName: string): boolean {
  const lower = fileName.toLowerCase();
  return SUPPORTED.some((ext) => lower.endsWith(ext));
}

export function isRenderable(fileName: string): boolean {
  return fileName.toLowerCase().endsWith(".pdf");
}

export async function extract(bytes: ArrayBuffer, fileName: string): Promise<Extraction> {
  const lower = fileName.toLowerCase();

  if (lower.endsWith(".pdf")) return extractPdf(bytes);
  if (lower.endsWith(".pptx")) return extractPptx(bytes);
  if (lower.endsWith(".docx")) return extractDocx(bytes);
  if (lower.endsWith(".html") || lower.endsWith(".htm")) return extractHtml(bytes);
  if (lower.endsWith(".txt") || lower.endsWith(".md") || lower.endsWith(".csv")) {
    return extractText(bytes);
  }

  throw new ExtractionError(
    `Cannot extract text from ${fileName}. Supported: PDF, PPTX, DOCX, HTML, TXT, MD, CSV.`,
  );
}

// --- PDF --------------------------------------------------------------------

let pdfjsPromise: Promise<typeof import("pdfjs-dist")> | null = null;

async function getPdfjs() {
  if (!pdfjsPromise) {
    // The LEGACY build: it avoids the newest syntax and is the one that
    // survives an extension's CSP without eval.
    pdfjsPromise = import("pdfjs-dist/legacy/build/pdf.mjs").then((lib) => {
      // Bundled, not fetched — MV3 will not load a worker from a CDN. The file
      // is deliberately named .js: Chrome serves extension files by extension
      // and rejects a module worker fetched from a .mjs URL on MIME grounds,
      // which silently broke EVERY PDF while DOCX kept working.
      lib.GlobalWorkerOptions.workerSrc = chrome.runtime.getURL("vendor/pdf.worker.js");
      return lib as unknown as typeof import("pdfjs-dist");
    });
  }
  return pdfjsPromise;
}

/**
 * Open a document, falling back to main-thread parsing if the worker will not
 * start.
 *
 * Slower, but a whole course of PDFs failing is a much worse outcome than a
 * sync that takes longer. `isEvalSupported: false` is required either way:
 * an extension's CSP forbids eval, and pdf.js otherwise reaches for it.
 */
async function openPdf(bytes: ArrayBuffer) {
  const pdfjs = await getPdfjs();
  const common = { isEvalSupported: false, useWorkerFetch: false, useSystemFonts: false };

  // The LOADING TASK owns destroy(), not the document proxy. Calling
  // doc.destroy() throws "t.destroy is not a function" after minification —
  // which failed every single PDF in a course while DOCX kept working.
  // TypeScript flagged exactly this and an earlier version cast the error
  // away; the cast was the bug.
  let task = pdfjs.getDocument({ data: new Uint8Array(bytes.slice(0)), ...common });
  try {
    return { doc: await task.promise, task };
  } catch (err) {
    console.warn("pdf.js worker unavailable, retrying on the main thread", err);
    await task.destroy().catch(() => {});
    task = pdfjs.getDocument({
      data: new Uint8Array(bytes.slice(0)),
      ...common,
      disableWorker: true,
    } as Parameters<typeof pdfjs.getDocument>[0]);
    return { doc: await task.promise, task };
  }
}

async function extractPdf(bytes: ArrayBuffer): Promise<Extraction> {
  // pdf.js takes ownership of the buffer, so it gets a copy — the caller still
  // needs the original to hash.
  const { doc, task } = await openPdf(bytes);

  const segments: Segment[] = [];
  try {
    for (let page = 1; page <= doc.numPages; page++) {
      const p = await doc.getPage(page);
      const content = await p.getTextContent();
      const text = content.items
        .map((item) => ("str" in item ? item.str : ""))
        .join(" ")
        .replace(/\s+/g, " ")
        .trim();
      if (text) segments.push({ position: `p.${page}`, text });
      p.cleanup();
    }
  } finally {
    await task.destroy().catch(() => {});
  }

  const total = segments.reduce((n, s) => n + s.text.length, 0);
  // A multi-page document yielding almost nothing is a scan.
  if (doc.numPages > 1 && total < 100) {
    return {
      segments,
      pageCount: doc.numPages,
      quality: "poor",
      reason:
        `Only ${total} characters of text across ${doc.numPages} pages — this looks ` +
        `like scanned images. OCR is not supported, so this file is not searchable.`,
    };
  }

  return { segments, pageCount: doc.numPages, quality: "ok" };
}

/**
 * Render one page so a diagram can actually be looked at.
 *
 * Returns raw base64 rather than a data URL, because the caller sends it to
 * the model as an inlineData part. Dimensions are capped: an uncapped A4 page
 * at scale 3 is a multi-megabyte PNG, and every megabyte is roughly 350k
 * tokens of quota.
 */
const MAX_RENDER_PX = 1400;

export async function renderPdfPage(
  bytes: ArrayBuffer,
  pageNumber: number,
  scale = 1.5,
): Promise<{
  base64: string;
  mimeType: string;
  width: number;
  height: number;
  pageCount: number;
}> {

  const { doc, task } = await openPdf(bytes);

  try {
    if (pageNumber < 1 || pageNumber > doc.numPages) {
      throw new ExtractionError(
        `Page ${pageNumber} is out of range; this document has ${doc.numPages}.`,
      );
    }
    const page = await doc.getPage(pageNumber);

    // Cap the long edge before rendering rather than downscaling after: a
    // 3000px canvas costs the memory whether or not we keep the pixels.
    const base = page.getViewport({ scale: 1 });
    const longEdge = Math.max(base.width, base.height);
    const capped = Math.min(scale, MAX_RENDER_PX / longEdge);
    const viewport = page.getViewport({ scale: Math.max(0.5, capped) });

    const canvas = new OffscreenCanvas(viewport.width, viewport.height);
    const context = canvas.getContext("2d") as unknown as CanvasRenderingContext2D;

    // White background: PDF pages are transparent, and JPEG has no alpha, so
    // without this the text renders on black.
    context.fillStyle = "#ffffff";
    context.fillRect(0, 0, viewport.width, viewport.height);

    await page.render({
      canvas: canvas as unknown as HTMLCanvasElement,
      canvasContext: context,
      viewport,
    }).promise;

    // JPEG, not PNG. A text-heavy page is 4-8x smaller as JPEG, and the model
    // is reading a diagram, not inspecting pixels.
    const blob = await canvas.convertToBlob({ type: "image/jpeg", quality: 0.82 });
    return {
      base64: await blobToBase64(blob),
      mimeType: "image/jpeg",
      width: viewport.width,
      height: viewport.height,
      pageCount: doc.numPages,
    };
  } finally {
    await task.destroy().catch(() => {});
  }
}

function blobToBase64(blob: Blob): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => {
      const result = String(reader.result);
      resolve(result.slice(result.indexOf(",") + 1));
    };
    reader.onerror = () => reject(reader.error);
    reader.readAsDataURL(blob);
  });
}

// --- OOXML (PPTX / DOCX) ----------------------------------------------------

/**
 * Created on demand, never at module scope. A service worker has no DOM, so a
 * top-level `new DOMParser()` makes the whole module unimportable there — and
 * the worker imports the tool registry.
 */
let _parser: DOMParser | null = null;
function getParser(): DOMParser {
  if (!_parser) _parser = new DOMParser();
  return _parser;
}

/** Pull every <a:t> / <w:t> text node in document order. */
function ooxmlText(xml: string, tag: string): string {
  const doc = getParser().parseFromString(xml, "application/xml");
  const nodes = doc.getElementsByTagName(tag);
  const parts: string[] = [];
  for (let i = 0; i < nodes.length; i++) {
    const t = nodes[i]?.textContent ?? "";
    if (t) parts.push(t);
  }
  return parts.join(" ").replace(/\s+/g, " ").trim();
}

async function extractPptx(bytes: ArrayBuffer): Promise<Extraction> {
  const zip = await JSZip.loadAsync(bytes);
  const segments: Segment[] = [];

  const slideNames = Object.keys(zip.files)
    .filter((n) => /^ppt\/slides\/slide\d+\.xml$/.test(n))
    .sort((a, b) => slideNumber(a) - slideNumber(b));

  for (const name of slideNames) {
    const n = slideNumber(name);
    const body = ooxmlText(await zip.files[name]!.async("string"), "a:t");
    if (body) segments.push({ position: `slide ${n}`, text: body });

    // Speaker notes carry the actual explanation surprisingly often — the
    // slide says "Amortized Analysis" and the notes explain it.
    const notesFile = zip.files[`ppt/notesSlides/notesSlide${n}.xml`];
    if (notesFile) {
      const notes = ooxmlText(await notesFile.async("string"), "a:t");
      // Notes slides repeat the slide number as a text run; drop that noise.
      if (notes && notes !== String(n)) {
        segments.push({ position: `slide ${n}`, text: notes, kind: "notes" });
      }
    }
  }

  const total = segments.reduce((acc, s) => acc + s.text.length, 0);
  if (slideNames.length > 1 && total < 100) {
    return {
      segments,
      pageCount: slideNames.length,
      quality: "poor",
      reason: `Almost no text across ${slideNames.length} slides — likely an image-only deck.`,
    };
  }

  return { segments, pageCount: slideNames.length, quality: "ok" };
}

const slideNumber = (name: string): number => Number(name.match(/(\d+)\.xml$/)?.[1] ?? 0);

async function extractDocx(bytes: ArrayBuffer): Promise<Extraction> {
  const zip = await JSZip.loadAsync(bytes);
  const file = zip.files["word/document.xml"];
  if (!file) throw new ExtractionError("This DOCX has no word/document.xml.");

  const doc = getParser().parseFromString(await file.async("string"), "application/xml");
  const paragraphs = doc.getElementsByTagName("w:p");
  const segments: Segment[] = [];
  let heading = "start";

  for (let i = 0; i < paragraphs.length; i++) {
    const p = paragraphs[i]!;
    const runs = p.getElementsByTagName("w:t");
    const parts: string[] = [];
    for (let j = 0; j < runs.length; j++) parts.push(runs[j]?.textContent ?? "");
    const text = parts.join("").replace(/\s+/g, " ").trim();
    if (!text) continue;

    // Heading styles are what make structure-aware chunking possible in a
    // DOCX: an outline's "Late Policy" section should stay in one chunk.
    const style = p.getElementsByTagName("w:pStyle")[0]?.getAttribute("w:val") ?? "";
    if (/^heading/i.test(style)) {
      heading = text;
      segments.push({ position: `§ ${text}`, text, kind: "heading" });
    } else {
      segments.push({ position: `§ ${heading}`, text });
    }
  }

  if (!segments.length) {
    return { segments, pageCount: null, quality: "poor", reason: "No text found in this DOCX." };
  }
  return { segments, pageCount: null, quality: "ok" };
}

// --- HTML / text ------------------------------------------------------------

function decode(bytes: ArrayBuffer): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(bytes);
}

async function extractHtml(bytes: ArrayBuffer): Promise<Extraction> {
  const doc = getParser().parseFromString(decode(bytes), "text/html");
  for (const el of [...doc.querySelectorAll("script, style, nav, noscript")]) el.remove();

  const segments: Segment[] = [];
  let heading = "start";

  for (const el of [...doc.body.querySelectorAll("h1, h2, h3, h4, p, li, td, pre")]) {
    const text = (el.textContent ?? "").replace(/\s+/g, " ").trim();
    if (!text) continue;
    if (/^h[1-4]$/i.test(el.tagName)) {
      heading = text;
      segments.push({ position: `§ ${text}`, text, kind: "heading" });
    } else {
      segments.push({ position: `§ ${heading}`, text });
    }
  }

  if (!segments.length) {
    return { segments, pageCount: null, quality: "poor", reason: "No readable text in this page." };
  }
  return { segments, pageCount: null, quality: "ok" };
}

async function extractText(bytes: ArrayBuffer): Promise<Extraction> {
  const text = decode(bytes);
  const segments: Segment[] = text
    .split(/\n{2,}/)
    .map((block, i) => ({ position: `¶ ${i + 1}`, text: block.replace(/\s+/g, " ").trim() }))
    .filter((s) => s.text);

  if (!segments.length) {
    return { segments, pageCount: null, quality: "poor", reason: "File is empty." };
  }
  return { segments, pageCount: null, quality: "ok" };
}
